content/blog/LFX-Mentorship-Blog.md

```diff
@@ -0,0 +1,425 @@
+
+[Rust programming language community server](https://discord.gg/rust-lang-community)
+
+[Firepit](https://discord.gg/4BUXJEB2)
+
+### My project
+
+My task during the internship was to add support for the Google Cloud connectors in Tremor. I worked on adding the Google Cloud Storage and Google Cloud Pub/Sub connectors. A detailed description of it can be found here:
+
+[https://github.com/tremor-rs/tremor-runtime/issues/724](https://github.com/tremor-rs/tremor-runtime/issues/724)
+
+Before explaining more about it, here's what a connector involves:
+
+**Sources:** Ingest data from the outside world and deserialise to events ( onramps )
+
+**Sinks:** Send serialised events to the outside world ( offramps )
+
+**Pipelines:** Business logic compiles to an event flow Directed Acyclic Graph
+
+![Connector](/img/blog/LFX-Mentorship/connector.png?raw=true)
+
+#### Google Cloud Storage connector:
+
+I wrote a GCS sink that can issue the basic GCS operations such as list buckets and objects, create, insert and delete objects from the Google Cloud Platform cloud storage service. The docs can be found here: [GCS offramp docs](https://docs.tremor.rs/artefacts/offramps/#gcs)
+
+#### Google Cloud Pub/Sub connector:
+
+The sink can issue the operation of sending a message to a Google Cloud Pub/Sub topic. It also allows creating a subscription to a topic to receive messages in it, with the option to enable/disable message ordering. The source allows receiving messages via a subscription in batches as well as one after another. The docs can be found here:
+[gpub offramp](https://docs.tremor.rs/artefacts/offramps/#gpub) and [gsub onramp](https://docs.tremor.rs/artefacts/onramps/#gsub)
+
+## Walk-Through Guide
+
+To get started, you need a service account on GCP and you will need a GCP pem file for certificates authentication and a service token json file.
+
+The command used to get the service_account `json` file:
+```bash
+gcloud iam service-accounts keys create key-file.json -iam-account=<iam-account-name>@<project-id>.iam.gserviceaccount.com
+```
+or 
+```bash
+gcloud iam service-accounts keys create key-file.json -iam-account=<mail-id-of-service-account>
+```
+
+Go to GCP dashboard → IAM & Admin → Service Accounts and get the email-id mentioned `<mail-id-of-service-account> = email-id`
+
+### Google Cloud Storage
+
+The following files are required:
+
+**outbound.trickle**
+
+```trickle
+select event from in into out
+```
+
+**inbound.trickle**
+
+```trickle
+select event from in into out
+```
+**test.yaml**
+
+```yaml
+offramp:
+  - id: stdout
+    type: stdout
+    codec: json
+    config:
+  - id: gcs
+    type: gcs
+    codec: json
+    postprocessors:
+      - gzip    
+    preprocessors:
+      - gzip
+    linked: true
+    config:
+      pem: <path-to-pem-file>
+onramp:
+  - id: stdin
+    type: file
+    codec: json
+    config:
+      source: /dev/stdin   
```

Comment:
Will this work without the lines preprocess or?

Reply:
how would the source know when an event ends without a lines pre processor?

Reply:
In the tests above the writer and reader or producer/consumer were the same application - so the configuration is a known safe quantity assuming a single writer ( as above ). Given the article's context as an experience report I think this is fine

Reply:
Nah the problem is that stdin is a streaming source so without a postprocessor to define framing it might just chunk a message in the middle. This needs a stdin preprocessor it will randomly not work for people trying to replicate it.

Reply:
🤦 I'm stupid, the file source is in fact line separated ... I'm really sorry for the confusion :/